File: flowquery-py/src/graph/data_resolver.py
"""Resolves pattern elements to data from the Database registry."""

from __future__ import annotations

from typing import Any, AsyncIterator, Dict, List, Optional, Union

from .data_cache import DataCache
from .database import Database
from .node import Node
from .node_data import NodeData
from .node_reference import NodeReference
from .physical_relationship import PhysicalRelationship
from .relationship import Relationship
from .relationship_data import RelationshipData


class DataResolver:
    """Resolves pattern elements (nodes and relationships) to data by querying
    the Database registry. Handles label compatibility, caching, and schema
    introspection — keeping query-resolution concerns separate from storage."""

    _instance: Optional['DataResolver'] = None

    def __init__(self) -> None:
        self._data_cache: DataCache = DataCache()

    @classmethod
    def get_instance(cls) -> 'DataResolver':
        if cls._instance is None:
            cls._instance = DataResolver()
        return cls._instance

    @property
    def data_cache(self) -> DataCache:
        return self._data_cache

    @data_cache.setter
    def data_cache(self, cache: DataCache) -> None:
        """Sets the data cache for the current query execution.
        Each top-level Runner creates its own DataCache instance."""
        self._data_cache = cache

    async def schema(self) -> List[Dict[str, Any]]:
        """Returns the graph schema with node/relationship labels and sample data."""
        return [item async for item in self._schema()]

    async def _schema(self) -> AsyncIterator[Dict[str, Any]]:
        """Async generator for graph schema."""
        db = Database.get_instance()
        for label, physical_node in db.nodes.items():
            records = await physical_node.data()
            entry: Dict[str, Any] = {"kind": "Node", "label": label}
            if records:
                sample = {k: v for k, v in records[0].items() if k != "id"}
                properties = list(sample.keys())
                if properties:
                    entry["properties"] = properties
                    entry["sample"] = sample
            yield entry

        for rel_type, type_map in db.relationships.items():
            for physical_rel in type_map.values():
                records = await physical_rel.data()
                entry_rel: Dict[str, Any] = {
                    "kind": "Relationship",
                    "type": rel_type,
                    "from_label": physical_rel.source.label if physical_rel.source else None,
                    "to_label": physical_rel.target.label if physical_rel.target else None,
                }
                if records:
                    sample = {
                        k: v for k, v in records[0].items()
                        if k not in ("left_id", "right_id")
                    }
                    properties = list(sample.keys())
                    if properties:
                        entry_rel["properties"] = properties
                        entry_rel["sample"] = sample
                yield entry_rel

    async def get_data(
        self, element: Union['Node', 'Relationship']
    ) -> Union['NodeData', 'RelationshipData']:
        """Gets data for a node or relationship."""
        db = Database.get_instance()
        if isinstance(element, Node):
            args = DataResolver._extract_args(element.properties)
            if len(element.labels) == 0:
                all_records = []
                for label, physical in db.nodes.items():
                    data = await self._data_cache.get(f"node:{label}", physical, None)
                    for record in data:
                        all_records.append({**record, "_label": label})
                return NodeData(all_records)
            if len(element.labels) > 1:
                all_records = []
                for lbl in element.labels:
                    phys_node = db.nodes.get(lbl)
                    if phys_node:
                        data = await self._data_cache.get(f"node:{lbl}", phys_node, args)
                        for record in data:
                            all_records.append({**record, "_label": lbl})
                return NodeData(all_records)
            node = db.get_node(element)
            if node is None:
                raise ValueError(f"Physical node not found for label {element.label}")
            data = await self._data_cache.get(f"node:{element.label}", node, args)
            label = element.label or ""
            records = [{**record, "_label": label} for record in data]
            return NodeData(records)
        elif isinstance(element, Relationship):
            args = DataResolver._extract_args(element.properties)
            entries = self._get_relationship_entries(element, db)
            if not entries:
                if len(element.types) == 0:
                    return RelationshipData([])
                suffix = "s" if len(element.types) > 1 else ""
                types = ", ".join(element.types)
                raise ValueError(
                    f"No physical relationships found for type{suffix} {types}"
                )
            all_records = []
            for type_name, phys_rel in entries:
                src = phys_rel.source.label if phys_rel.source else None
                tgt = phys_rel.target.label if phys_rel.target else None
                cache_key = f"rel:{src or ''}:{type_name}:{tgt or ''}"
                records = await self._data_cache.get(cache_key, phys_rel, args)
                for record in records:
                    all_records.append({**record, "_type": type_name})
            return RelationshipData(all_records)
        else:
            raise ValueError("Element is neither Node nor Relationship")

    @staticmethod
    def _resolve_labels(node: Optional['Node']) -> List[str]:
        """Resolve labels, following NodeReference if the node itself has none."""
        if node is None:
            return []
        if node.labels:
            return node.labels
        if isinstance(node, NodeReference):
            ref = node.reference
            if isinstance(ref, Node):
                return list(ref.labels)
        return []

    @staticmethod
    def _is_relationship_compatible(
        relationship: 'Relationship', physical: 'PhysicalRelationship'
    ) -> bool:
        """Check if a MATCH pattern's endpoint labels are compatible."""
        def match(pattern: List[str], label: Optional[str]) -> bool:
            return len(pattern) == 0 or (label is not None and label in pattern)

        src_labels = DataResolver._resolve_labels(relationship.source)
        tgt_labels = DataResolver._resolve_labels(relationship.target)
        phys_src = physical.source.label if physical.source else None
        phys_tgt = physical.target.label if physical.target else None
        if relationship.direction == "left":
            return match(src_labels, phys_tgt) and match(tgt_labels, phys_src)
        return match(src_labels, phys_src) and match(tgt_labels, phys_tgt)

    def _get_relationship_entries(
        self,
        relationship: 'Relationship',
        db: Database,
    ) -> list[tuple[str, 'PhysicalRelationship']]:
        """Collect physical relationships matching the pattern's types and endpoint labels."""
        types = relationship.types or list(db.relationships.keys())
        return [
            (t, p)
            for t in types if t in db.relationships
            for p in db.relationships[t].values()
            if DataResolver._is_relationship_compatible(relationship, p)
        ]

    @staticmethod
    def _extract_args(
        properties: Dict[str, Any],
    ) -> Optional[Dict[str, Any]]:
        """Extracts property constraint values to pass as $args to the inner query."""
        if not properties:
            return None
        return {key: expression.value() for key, expression in properties.items()}

File: src/graph/data_resolver.ts
import DataCache from "./data_cache";
import Database from "./database";
import Node from "./node";
import NodeData, { NodeRecord } from "./node_data";
import NodeReference from "./node_reference";
import PhysicalRelationship from "./physical_relationship";
import Relationship from "./relationship";
import RelationshipData, { RelationshipRecord } from "./relationship_data";
import { attachVirtualSource, getVirtualSource } from "./virtual_sources";

/**
 * Resolves pattern elements (nodes and relationships) to data by querying
 * the Database registry. Handles label compatibility, caching, and schema
 * introspection — keeping query-resolution concerns separate from storage.
 */
class DataResolver {
    private static instance: DataResolver;
    private _dataCache: DataCache = new DataCache();

    public static getInstance(): DataResolver {
        if (!DataResolver.instance) {
            DataResolver.instance = new DataResolver();
        }
        return DataResolver.instance;
    }
    /**
     * Sets the data cache for the current query execution.
     * Each top-level Runner creates its own DataCache instance.
     */
    public set dataCache(cache: DataCache) {
        this._dataCache = cache;
    }

    public async schema(): Promise<Record<string, any>[]> {
        const db = Database.getInstance();
        const result: Record<string, any>[] = [];

        for (const [label, physical] of db.nodes) {
            const records = await physical.data();
            const entry: Record<string, any> = { kind: "Node", label };
            if (records.length > 0) {
                const { id, ...sample } = records[0];
                const properties = Object.keys(sample);
                if (properties.length > 0) {
                    entry.properties = properties;
                    entry.sample = sample;
                }
            }
            result.push(entry);
        }

        for (const [type, typeMap] of db.relationships) {
            for (const physical of typeMap.values()) {
                const records = await physical.data();
                const entry: Record<string, any> = {
                    kind: "Relationship",
                    type,
                    from_label: physical.source?.label || null,
                    to_label: physical.target?.label || null,
                };
                if (records.length > 0) {
                    const { left_id, right_id, ...sample } = records[0];
                    const properties = Object.keys(sample);
                    if (properties.length > 0) {
                        entry.properties = properties;
                        entry.sample = sample;
                    }
                }
                result.push(entry);
            }
        }

        return result;
    }

    public async getData(element: Node | Relationship): Promise<NodeData | RelationshipData> {
        const db = Database.getInstance();
        if (element instanceof Node) {
            const args = DataResolver.extractArgs(element.properties);
            if (element.labels.length === 0) {
                // Unlabeled node: match all physical nodes in the database
                const allRecords: NodeRecord[] = [];
                for (const [label, physical] of db.nodes) {
                    const data = await this._dataCache.get(`node:${label}`, physical, null);
                    for (const record of data as NodeRecord[]) {
                        const enriched = { ...record, _label: label };
                        const src = getVirtualSource(record);
                        if (src !== undefined) attachVirtualSource(enriched, src);
                        allRecords.push(enriched);
                    }
                }
                return new NodeData(allRecords);
            }
            if (element.labels.length > 1) {
                // ORed labels: collect from all matching physical nodes
                const allRecords: NodeRecord[] = [];
                for (const lbl of element.labels) {
                    const physical = db.nodes.get(lbl);
                    if (physical) {
                        const data = await this._dataCache.get(`node:${lbl}`, physical, args);
                        for (const record of data as NodeRecord[]) {
                            const enriched = { ...record, _label: lbl };
                            const src = getVirtualSource(record);
                            if (src !== undefined) attachVirtualSource(enriched, src);
                            allRecords.push(enriched);
                        }
                    }
                }
                return new NodeData(allRecords);
            }
            const node = db.getNode(element);
            if (node === null) {
                throw new Error(`Physical node not found for label ${element.label}`);
            }
            const data = await this._dataCache.get(`node:${element.label}`, node, args);
            const label = element.label;
            const records = (data as NodeRecord[]).map((record) => {
                const enriched = { ...record, _label: label };
                const src = getVirtualSource(record);
                if (src !== undefined) attachVirtualSource(enriched, src);
                return enriched;
            });
            return new NodeData(records);
        } else if (element instanceof Relationship) {
            const args = DataResolver.extractArgs(element.properties);
            const entries = this.getRelationshipEntries(element, db);
            if (entries.length === 0) {
                if (element.types.length === 0) {
                    return new RelationshipData([]);
                }
                throw new Error(
                    `No physical relationships found for type${element.types.length > 1 ? "s" : ""} ${element.types.join(", ")}`
                );
            }
            const allRecords: RelationshipRecord[] = [];
            for (const [typeName, physical] of entries) {
                const cacheKey = `rel:${physical.source?.label ?? ""}:${typeName}:${physical.target?.label ?? ""}`;
                const records = (await this._dataCache.get(
                    cacheKey,
                    physical,
                    args
                )) as RelationshipRecord[];
                for (const record of records) {
                    const enriched = { ...record, _type: typeName };
                    const src = getVirtualSource(record);
                    if (src !== undefined) attachVirtualSource(enriched, src);
                    allRecords.push(enriched);
                }
            }
            return new RelationshipData(allRecords);
        } else {
            throw new Error("Element is neither Node nor Relationship");
        }
    }

    /** Resolve labels from a node, following NodeReference if needed. */
    private static resolveLabels(node: Node | null): string[] {
        if (!node) return [];
        if (node.labels.length > 0) return node.labels;
        if (node instanceof NodeReference) {
            const ref = node.reference;
            if (ref instanceof Node) return ref.labels;
        }
        return [];
    }

    /**
     * Check whether a MATCH pattern's endpoint labels are compatible with
     * a stored physical relationship's source/target labels.
     */
    private static isRelationshipCompatible(
        relationship: Relationship,
        physical: PhysicalRelationship
    ): boolean {
        const match = (pattern: string[], label: string | null): boolean =>
            pattern.length === 0 || (label !== null && pattern.includes(label));
        const srcLabels = DataResolver.resolveLabels(relationship.source);
        const tgtLabels = DataResolver.resolveLabels(relationship.target);
        const physSrc = physical.source?.label ?? null;
        const physTgt = physical.target?.label ?? null;
        return relationship.direction === "left"
            ? match(srcLabels, physTgt) && match(tgtLabels, physSrc)
            : match(srcLabels, physSrc) && match(tgtLabels, physTgt);
    }

    /**
     * Collect all physical relationships matching the pattern's types and endpoint labels.
     * If types is empty, all registered relationship types are considered.
     */
    private getRelationshipEntries(
        relationship: Relationship,
        db: Database
    ): [string, PhysicalRelationship][] {
        const types =
            relationship.types.length === 0
                ? Array.from(db.relationships.keys())
                : relationship.types;

        const result: [string, PhysicalRelationship][] = [];
        for (const type of types) {
            const typeMap = db.relationships.get(type);
            if (!typeMap) continue;
            for (const physical of typeMap.values()) {
                if (DataResolver.isRelationshipCompatible(relationship, physical)) {
                    result.push([type, physical]);
                }
            }
        }
        return result;
    }

    private static extractArgs(properties: Map<string, any>): Record<string, any> | null {
        if (properties.size === 0) {
            return null;
        }
        const args: Record<string, any> = {};
        for (const [key, expression] of properties) {
            args[key] = expression.value();
        }
        return args;
    }
}

export default DataResolver;
